### PR TITLE
post: what is ng ivy

### PR DESCRIPTION
**Pull request by cexbrayat**
I initially started to write about Angular 8.0,
but Ivy itself is too big, so it now has its own article.
I plan to release it in a week or two (depending on ng-conf this week).

I hope this is clear enough, let me know!

### CODE REVIEW

File: _posts/2019-05-06-what-is-angular-ivy.md [hunk 1]
```diff
@@ -0,0 +1,362 @@
...
+The thing a lot of developers don't really know is that
+this HTML will never touch the browser.
+It will be compiled by Angular into JavaScript instructions,
+to create the appropriate DOM when the component appears on the page,
+and to update the component when its state changes.
+That's why a big part of Angular is its compiler:
+it takes all your HTML and generates the JS code necessary.
+This compiler (and the runtime) has been completely rewritten over the last year,
+and this is what Ivy is about.
+Thi is not the first rewrite:
```
Comment (clacote) [inline _posts/2019-05-06-what-is-angular-ivy.md]: Thi -> This

File: _posts/2019-05-06-what-is-angular-ivy.md [hunk 2]
```diff
@@ -0,0 +1,362 @@
...
+the Angular team is literally changing the engine (previously called View Engine) while driving.
+
+
+## The goals of Ivy
+
+Ivy is a very important stepping stone in the Angular history.
+It changes how the framework internally works,
+without changing how we write Angular applications.
+
+If the parallel makes some sense for you, it's very similar to React and the "Fiber rewrite".
```
Comment (clacote) [inline _posts/2019-05-06-what-is-angular-ivy.md]: makes more sense for you -> to you, no?

File: _posts/2019-05-06-what-is-angular-ivy.md [hunk 3]
```diff
@@ -0,0 +1,362 @@
...
+So when you'll update to Angular 8,
+the schematics will run and tweak a few things in your code to be "Ivy-ready"
+when the time comes.
+The plan is to enable Ivy by default in the future,
+probably in V9 or v10.
+
+
+## Differences in generated code
+
+Let's dive into the gritty details 🤓.
```
Comment (clacote) [inline _posts/2019-05-06-what-is-angular-ivy.md]: I don't know what `gritty` means, but maybe that's me, and this is not important.

File: _posts/2019-05-06-what-is-angular-ivy.md [hunk 4]
```diff
@@ -0,0 +1,362 @@
...
+
+But it also turns out that previously, when you were working on your application,
+Angular had to recompile everything inside your module to know what had changed,
+because the generated code of a component could be using internal details of another component.
+
+Now each component references the directives and components it uses only by their public APIs.
+So if you modify an internal detail of a component or a directive,
+only the actual components that use it will be recompiled!
+It could lead to huge benefits in rebuild times for applications with dozens of components and directives,
+as you will go to recompile all of them to recompile just the needed ones.
```
Comment (clacote) [inline _posts/2019-05-06-what-is-angular-ivy.md]: you will go to recompile all -> you will go _from_ recompil_ing_ all, no?

File: _posts/2019-05-06-what-is-angular-ivy.md [hunk 5]
```diff
@@ -0,0 +1,362 @@
...
+
+Let's take an example with our `PonyComponent` that declares its input like this:
+
+    @Input('pony) ponyModel: PonyModel;
```
Comment (clacote) [inline _posts/2019-05-06-what-is-angular-ivy.md]: Missing final quote after `pony`

File: _posts/2019-05-06-what-is-angular-ivy.md [hunk 6]
```diff
@@ -0,0 +1,362 @@
...
+
+A little trivia of Angular history:
+modules were introduced fairly lately in Angular development,
+just a few months before the stable release.
+Previously, during the 2.0 beta phase,
+you had to reference manually each component and directive used
+in a component directly in its decorator.
+Modules were introduced to avoid that, but the downside was that it become the smallest unit of compilation:
+changing one element of the module lead to recompile all the elements of the module.
+You can see how the code generated in Ivy take us back to what was originally designed by the Angular team,
```
Comment (clacote) [inline _posts/2019-05-06-what-is-angular-ivy.md]: take -> takes

File: _posts/2019-05-06-what-is-angular-ivy.md [hunk 7]
```diff
@@ -0,0 +1,362 @@
...
+with a `directives` property generatedin the `ngComponentDef` field .
```
Comment (clacote) [inline _posts/2019-05-06-what-is-angular-ivy.md]: generatedin -> generated in

Comment (clacote) [reply _posts/2019-05-06-what-is-angular-ivy.md]: useless space before final `.`

File: _posts/2019-05-06-what-is-angular-ivy.md [hunk 8]
```diff
@@ -0,0 +1,362 @@
...
+I also heard the team talk about changing/removing modules one day in various conferences,
```
Comment (clacote) [inline _posts/2019-05-06-what-is-angular-ivy.md]: talk -> talking

File: _posts/2019-05-06-what-is-angular-ivy.md [hunk 9]
```diff
@@ -0,0 +1,362 @@
...
+
+That's why the team expects big improvements on the size of small applications,
+and especially on a Hello World application (which previously produced a big bundle for Hello World),
+or for an [Angular Element](/2018/05/29/angular-elements/).
+
+For the medium to large applications,
+the situation should not change much with the first release of Ivy.
+The bundles should be roughly the same sizes (or even slightly bigger) as they are with View Engine.
+The Angular team will have time to focus on that once they are sure that there is no regression with Ivy,
+and we can hope for smaller bundles in every cases in the future.
```
Comment (clacote) [inline _posts/2019-05-06-what-is-angular-ivy.md]: every cases -> every case, no?

File: _posts/2019-05-06-what-is-angular-ivy.md [hunk 10]
```diff
@@ -0,0 +1,362 @@
...
+
+
+## Runtime performances
+
+Ivy has no particular focus on performances, at least not in the first release.
+It has been designed to be very efficient memory-wise, some mecanics have been improved,
```
Comment (clacote) [inline _posts/2019-05-06-what-is-angular-ivy.md]: mecanics -> mechanics

File: _posts/2019-05-06-what-is-angular-ivy.md [hunk 11]
```diff
@@ -0,0 +1,362 @@
...
+and it's still designed to avoid mega-morphic calls,
```
Comment (clacote) [inline _posts/2019-05-06-what-is-angular-ivy.md]: a link to your blog post about mega-morphism?

File: _posts/2019-05-06-what-is-angular-ivy.md [hunk 12]
```diff
@@ -0,0 +1,362 @@
...
+## Retro-compatibility
+
+I explained that the Ivy compiler takes the decorators in your TypeScript code,
+and then generates a static field in the class.
+But the current libraries shipped on NPM don't have their decorator anymore,
+they usually ship the JavaScript code resulting from the compilation.
+And Ivy needs these static fields to properly work,
+so are we stuck until every library we use ship the new version?
+
+Hopefuly no, we aren't 😊.
```
Comment (clacote) [inline _posts/2019-05-06-what-is-angular-ivy.md]: Hopefuly -> Hopefully

Comment (clacote) [reply _posts/2019-05-06-what-is-angular-ivy.md]: 😊 -> 😅 (my favorite)

File: _posts/2019-05-06-what-is-angular-ivy.md (hunk 11 above)
Comment (cexbrayat) [reply _posts/2019-05-06-what-is-angular-ivy.md]: i don't think we have one, only some references in the ebook

File: _posts/2019-05-06-what-is-angular-ivy.md [hunk 13]
```diff
@@ -0,0 +1,362 @@
...
+## Better template type checking
+
+The Angular compiler has an option which is often overlooked in my opinion: `fullTemplateTypeCheck`.
+When enabled, the compiler tries to analyze the templates more deeply.
+I showed some examples of what it's capable of,
+when it was introduced in [Angular 5](/2017/11/02/what-is-new-angular-5/).
+This option is now more powerful in Ivy,
+and will probably be even more powerful in the future.
+
+For example, one of the feature already available in Ivy
```
Comment (jnizet) [inline _posts/2019-05-06-what-is-angular-ivy.md]: of the features

File: _posts/2019-05-06-what-is-angular-ivy.md [hunk 14]
```diff
@@ -0,0 +1,362 @@
...
+is the type-checking of component and directive inputs.
+Imagine an `ImageComponent` with an input called `size`, of type `number`.
+If another component uses `ImageComponent` and try to pass a boolean value into the `size` property,
```
Comment (jnizet) [inline _posts/2019-05-06-what-is-angular-ivy.md]: and tries to pass

File: _posts/2019-05-06-what-is-angular-ivy.md [hunk 15]
```diff
@@ -0,0 +1,362 @@
...
+you'll see the error: `Type 'boolean' is not assignable to type 'number'.`.
+
+This is a just an example of what Ivy will be capable of,
+and these features are very interesting for large applications.
+
+
+## Retro-compatibility
```
Comment (jnizet) [inline _posts/2019-05-06-what-is-angular-ivy.md]: Backward compatibility

File: _posts/2019-05-06-what-is-angular-ivy.md [hunk 16]
```diff
@@ -0,0 +1,362 @@
...
+
+
+## Future possibilities
+
+Angular 8.0 is really the first step for Ivy.
+When it will be stable enough,
+it will become the default.
+And then the team can start working on adding other features more easily.
+Like the [i18n service](https://github.com/angular/angular/issues/11405),
+probably one of the most awaited feature.
```
Comment (jnizet) [inline _posts/2019-05-06-what-is-angular-ivy.md]: features

File: _posts/2019-05-06-what-is-angular-ivy.md [hunk 17]
```diff
@@ -0,0 +1,362 @@
...
+Or the possibility to have [metaprogramming or higher order components](https://blog.nrwl.io/metaprogramming-higher-order-components-and-mixins-with-angular-ivy-75748fcbc310).
+Or to lazy-load a single component instead of a module.
+Or to have JiT components and AoT components work with each other.
+Or to manually craft a template, by hand-writing the generated instructions
+to squeeze the best performances.
+And probably tons of other ideas the team has,
+and have not talked about yet.
+
+I hope this clarified a bit what Ivy is about,
+and that you'll give a try in the next months!
```
Comment (jnizet) [inline _posts/2019-05-06-what-is-angular-ivy.md]: you'l give it a try

**Pull request merged**